Infrastructure-Scripts/New-ParametersFile.ps1

```diff
@@ -0,0 +1,112 @@
+<#
+    .SYNOPSIS
+    Dynamically create a parameters.json file from an arm template
+
+    .DESCRIPTION
+    Dynamically create a parameters.json file with values from an arm template and environment variables matching the arm template parameters
+
+    .PARAMETER TemplateFilePath
+    File path to the ARM template
+
+    .PARAMETER ParametersFilePath
+    File path to store the generated Parameters
+
+    .EXAMPLE
+    New-ParametersFile.ps1 -TemplateFilePath "C:\template.json" -ParametersFilePath "C:\template.parameters.json"
+#>
+[CmdletBinding()]
+Param(
+    [Parameter(Mandatory = $true)]
+    [String]$TemplateFilePath,
+    [Parameter(Mandatory = $true)]
+    [String]$ParametersFilePath
+)
+
+try {
+
+    $TemplateParameters = (Get-Content -Path $TemplateFilePath -Raw | ConvertFrom-Json).Parameters
+    $ParametersFile = [PSCustomObject]@{
+        "`$schema"     = "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#"
+        contentVersion = "1.0.0.0"
+        parameters     = @{ }
+    }
+    $ParameterObjects = $TemplateParameters.PSObject.Members | Where-Object MemberType -eq NoteProperty
+
+    foreach ($ParameterObject in $ParameterObjects) {
+        $ParameterType = $ParameterObject.Value.Type
+        $ParameterName = $ParameterObject.Name
+        $ParameterValue = (Get-Item -Path "env:$ParameterName" -ErrorAction SilentlyContinue).Value
+
+        if (!$ParameterValue) {
+            Write-Verbose -Message "Environment variable for $ParameterName was not found, attempting default value"
+            $ParameterValue = $TemplateParameters.$ParameterName.defaultValue
```

Comment:
Does this need to be `$ParameterValue = $ParameterObject.defaultValue` ?

Reply:
The $ParameterObject returns an object similar to below: 
```Value           : @{type=string; defaultValue=String}
MemberType      : NoteProperty
IsSettable      : True
IsGettable      : True
TypeNameOfValue : System.Management.Automation.PSCustomObject
Name            : DefaultString
IsInstance      : True
```

The ParameterObjects are used to distinguish the Parameter Name and Parameter Type. 
It is possible to do `$ParameterValue = $ParameterObject.Value.defaultValue` but it goes back into the original arm template and searches for the specific parameter and its default value instead.

Reply:
OK, I think as you're already referencing the $ParameterObject within the foreach loop I think it makes sense to keep referencing that. Maybe it also has some performance benefit as it doesn't have to search through the entire parameter list again?

Reply:
Yep, does make more sense and good point on the performance benefit as well. Updated 👍